doc/source/tutorials/concurrency-consistency.rst

```diff
@@ -0,0 +1,304 @@
+Concurrency and Consistency
+===========================
+
+In addition to using parallelism internally, TileDB is designed
+having **parallel programming** in mind. Specifically,
+scientific computing users may be accustomed to using *multi-processing*
+(e.g., via `MPI <https://www.mpich.org/>`__ or
+`Dask <https://dask.pydata.org/en/latest/>`__),
+or writing *multi-threaded* programs to speed up performance.
+In this tutorial we explain
+how TileDB enables *concurrency* via **thread-/process-safe** queries.
+In addition, we include a discussion on *consistency*, i.e.,
+on how each thread/process views the contents of an array at any
+given moment when other threads/processes are working concurrently.
+
+We will initially describe concurrency and consistency for arrays.
+The case of key-value stores is similar, but there are a few nuances
+we stress out at the end of the tutorial. Note that TileDB object
+management and VFS operations are not currently thread-/process-safe,
+but we will consider providing this feature in a future version.
+
+Array concurrency
+-----------------
+
+Defining an array, i.e., storing the array schema on disk is neither
+thread- nor process-safe. We do not expect a practical scenario
+where multiple threads/processes attempt to define the same array
+in parallel. Moreover, defining the array is extremely lightweight
+(and parallelized internally), thus it will never be a performance
+bottleneck. Therefore, in this section we focus on concurrent
+queries.
+
+In TileDB, a *write operation* is the process of creating a write
+query object, submitting the query (potentially multiple times in
+the case of global writes) and finalizing the query object
+(important only in global writes). The code snippet shown below
+describes an **atomic write operation**, i.e., a block of functions
+that must be treated atomically by each thread. For example, do not
+attempt to prepare the
+query object with multiple threads, or have multiple threads
+invoke ``query.submit()`` for the same ``Query`` object ``query``.
+Instead, you can have multiple threads create *separate* ``Query`` objects
+(e.g., ``query_1``, ``query_2``) for the same array even sharing the
+same context, and prepare and submit them in parallel having each thread
+use the atomic block shown below.
+
+.. content-tabs::
+
+   .. tab-container:: cpp
+      :title: C++
+
+      .. code-block:: c++
+
+         // ### An atomic write operation ###
+         // Optionally: Array array(ctx, array_name, TILEDB_WRITE);
```

Comment:
How is this optional? The `Query` constructor needs an array object.

Reply:
It is optional inside the atomic write operation block. You can create/open the array outside the block and have multiple threads use the same `Array` object.